File: authorization/homomorphic_encryption.rs
// homomorphic_encryption.rs
use rust_homomorphic_encryption::{HomomorphicEncryption, PublicKey, Ciphertext};

struct SecureDataSharing {
    he: HomomorphicEncryption,
    public_key: PublicKey,
}

impl SecureDataSharing {
    fn new() -> Self {
        let he = HomomorphicEncryption::new();
        let public_key = he.generate_public_key();
        SecureDataSharing { he, public_key }
    }

    fn encrypt(&self, plaintext: &str) -> Ciphertext {
        self.he.encrypt(plaintext, &self.public_key)
    }

    fn decrypt(&self, ciphertext: &Ciphertext) -> String {
        self.he.decrypt(ciphertext, &self.public_key)
    }

    fn evaluate(&self, ciphertext1: &Ciphertext, ciphertext2: &Ciphertext) -> Ciphertext {
        self.he.evaluate(ciphertext1, ciphertext2, &self.public_key)
    }
}

File: projects/PiHE/crypto/homomorphic_encryption.py
import pyseal

class HomomorphicEncryption:
    def __init__(self, public_key, private_key):
        self.public_key = public_key
        self.private_key = private_key
        self.context = pyseal.SEALContext.Create()
        self.key_context = pyseal.KeyContext(self.context)
        self.encryptor = pyseal.Encryptor(self.context, self.public_key)
        self.decryptor = pyseal.Decryptor(self.context, self.private_key)
        self.evaluator = pyseal.Evaluator(self.context)

    def encrypt(self, plaintext):
        ciphertext = pyseal.Ciphertext()
        self.encryptor.encrypt(plaintext, ciphertext)
        return ciphertext

    def decrypt(self, ciphertext):
        plaintext = pyseal.Plaintext()
        self.decryptor.decrypt(ciphertext, plaintext)
        return plaintext

    def add(self, ciphertext1, ciphertext2):
        result = pyseal.Ciphertext()
        self.evaluator.add(ciphertext1, ciphertext2, result)
        return result

    def multiply(self, ciphertext1, ciphertext2):
        result = pyseal.Ciphertext()
        self.evaluator.multiply(ciphertext1, ciphertext2, result)
        return result
